### ISSUE TITLE
type not inferred correctly in generic function

### ISSUE THREAD

**Issue by yaymukund** [microsoft/TypeScript#32109]
**TypeScript Version:**  3.5.2

**Search Terms:** generic type constraint undefined enum

**Code**

```ts
enum MyEvent {
    Start = 'start',
    Stop = 'stop',
}

interface MyEventOptions {
    [MyEvent.Start]: { startTime: number },
    [MyEvent.Stop]: { endTime: number },
}

class EventService {
    private eventCounts: { [key in MyEvent]?: number };

    constructor() {
        // I don't want to have to list every possible enum member here
        this.eventCounts = {};
    }

    private startEvent() {
        this.report(MyEvent.Start, { startTime: 0 });

        // This should be invalid:
        // 
        // this.report(MyEvent.Start, { endTime: 0 });
    }

    private report<T extends keyof MyEventOptions>(event: T, options: MyEventOptions[T]) {
        // errors with `Type 'undefined' is not assignable to type 'number'.`
        const count: number = this.eventCounts[event] || 0;

        // This would do, for example:
        //
        // request(mixpanelUrl, { name: event, options });
    }
}
```

It doesn't complain if you change it to `const count: number = (this.eventCounts.get(event) as number | undefined) || 0;`

**Expected behavior:** Compiles without a problem.

**Actual behavior:**

```
Type '{ start?: number | undefined; stop?: number | undefined; }[T]' is not assignable to type 'number'.
  Type 'number | undefined' is not assignable to type 'number'.
    Type 'undefined' is not assignable to type 'number'.
```

**Playground Link:** https://is.gd/S57reZ

**Related Issues:** https://github.com/microsoft/TypeScript/issues/32017

**Comment by sandersn** [comment]
@yaymukund typescript@next behaves the way you want it to, as far as I can tell. Can you confirm?

**Comment by yaymukund** [comment]
@sandersn Thank you for taking a look.

It's not fixed even in `typescript@next`. It looks like it only errors when `strictNullChecks` is `true`. I've made [a repository that reproduces the error](https://github.com/yaymukund/typescript-issue-32109) in typescript@next.

**Comment by RyanCavanaugh** [comment]
@yaymukund see #31908

**Comment by yaymukund** [comment]
Thanks for the heads up— I'll keep an eye on that issue!
